### PR TITLE
Print apiVersion info when failed to execute kubectl apply -f XXX.yaml

### PR DESCRIPTION
**Pull request by chentao1596**
**What this PR does / why we need it**:

Fixes https://github.com/kubernetes/kubernetes/issues/55216

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by brendandburns** [comment]
Unit tests are failing.

Can you add examples of the output before and after your change to the PR description?

Thanks
--brendan

**Comment by chentao1596** [comment]
@brendanburns 
Hi, two examples:

The first, example 1:
kind-not-match.yaml 
```
apiVersion: scheduling.io/v1alpha 
kind: PriorityPreemptionClass
metadata:
  name: high-priority
value: 1000
globalDefault: false
description: "This priority class should be used for higher level app."
```
before:
```
# kubectl apply -f kind-not-match.yaml 
error: unable to recognize "kind-not-match.yaml": no matches for scheduling.io/, Kind=PriorityPreemptionClass
```
after:
```
# kubectl apply -f kind-not-match.yaml 
error: unable to recognize "kind-not-match.yaml": no matches for scheduling.io/v1alpha, Kind=PriorityPreemptionClass
```


------------------------------------------------------------------
Second, example 2:
kind-not-match-no-group.yaml 
```
apiVersion: v1 
kind: PriorityClass
metadata:
  name: high-priority
value: 1000
globalDefault: false
description: "This priority class should be used for higher level app."
```
before:
```
# kubectl apply -f kind-not-match-no-group.yaml 
error: unable to recognize "kind-not-match-no-group.yaml": no matches for /, Kind=PriorityClass
```
after:
```
# kubectl apply -f kind-not-match-no-group.yaml 
error: unable to recognize "kind-not-match-no-group.yaml": no matches for v1, Kind=PriorityClass
```

**Comment by chentao1596** [comment]
@brendanburns 
I'm going to try to find the errors in the unit tests, but so far, i am not sure why it failed, if you have some suggestion, please tell me, thank you, ^_^！

**Comment by brendandburns** [comment]
/retest

My guess is that something in the scripts is looking for that text, though I'm not certain.

**Comment by chentao1596** [comment]
@brendandburns Ok, I try it again, thank you, ^_^!

**Comment by chentao1596** [comment]
@brendandburns
Hi, i have finished it.

I think the error is caused by the other bug, the parameter to be passed should be "version", not "apiversion" when invoke the "restMapper.RESTMapping" in the "controller/garbagecollector".

thank you!

**Comment by brendandburns** [comment]
/lgtm
/approve

Thanks for your patience!

File: staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper.go [hunk 1]
```diff
@@ -179,7 +179,7 @@ func (m MultiRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (*
 	if len(errors) > 0 {
 		return nil, utilerrors.NewAggregate(errors)
 	}
-	return nil, &NoKindMatchError{PartialKind: gk.WithVersion("")}
+	return nil, &NoKindMatchError{PartialKind: gk.WithVersion(strings.Join(versions, "&"))}
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper.go]: joining multiple versions into a single string field like this isn't appropriate

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper.go]: we can't assume this is only used for display... PartialKind should not contain inaccurate information... if we want to include additional information for display (like all the versions we tried), let's do that in a well-structured way

File: staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper.go [hunk 2]
```diff
@@ -204,7 +204,7 @@ func (m MultiRESTMapper) RESTMappings(gk schema.GroupKind, versions ...string) (
 		return nil, utilerrors.NewAggregate(errors)
 	}
 	if len(allMappings) == 0 {
-		return nil, &NoKindMatchError{PartialKind: gk.WithVersion("")}
+		return nil, &NoKindMatchError{PartialKind: gk.WithVersion(strings.Join(versions, "&"))}
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper.go]: same comment here

File: staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go [hunk 3]
```diff
@@ -136,7 +136,11 @@ func (gvk GroupVersionKind) GroupVersion() GroupVersion {
 }
 
 func (gvk GroupVersionKind) String() string {
-	return gvk.Group + "/" + gvk.Version + ", Kind=" + gvk.Kind
+	if len(gvk.Group) == 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go]: revert this change, being able to tell the difference between version and group in the output is useful

File: staging/src/k8s.io/apimachinery/pkg/api/meta/priority.go [hunk 4]
```diff
@@ -153,7 +153,7 @@ func kindMatches(pattern schema.GroupVersionKind, kind schema.GroupVersionKind)
 }
 
 func (m PriorityRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (mapping *RESTMapping, err error) {
-	mappings, err := m.Delegate.RESTMappings(gk)
+	mappings, err := m.Delegate.RESTMappings(gk, versions...)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/priority.go]: @deads2k does passing these through look correct to you? is the underlying delegate intended to know about the user versions, or does PriorityRESTMapper take responsibility for making use of those versions?

File: pkg/controller/garbagecollector/operations.go [hunk 5]
```diff
@@ -34,7 +34,7 @@ import (
 // namespace> tuple to a unversioned.APIResource struct.
 func (gc *GarbageCollector) apiResource(apiVersion, kind string) (*metav1.APIResource, error) {
 	fqKind := schema.FromAPIVersionAndKind(apiVersion, kind)
-	mapping, err := gc.restMapper.RESTMapping(fqKind.GroupKind(), apiVersion)
+	mapping, err := gc.restMapper.RESTMapping(fqKind.GroupKind(), fqKind.Version)
```
Comment (liggitt) [inline pkg/controller/garbagecollector/operations.go]: I *think* the fact that GC is currently working, when we were passing in an invalid version here means that the current behavior of PriorityRESTMapper treats the specified versions as a hint, not a requirement.

Fixing this bug here to pass the right version is good, but I'm not as sure about the change at https://github.com/kubernetes/kubernetes/pull/55218/files#diff-1054b3cbb6002ca622530dd7b2733558R156

File: staging/src/k8s.io/apimachinery/pkg/api/meta/priority.go (hunk 4 above)
Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/priority.go]: > @deads2k does passing these through look correct to you? is the underlying delegate intended to know about the user versions, or does PriorityRESTMapper take responsibility for making use of those versions?

They should be passed through.  With our current chain, I think the end result came out the same, but each constituent should get its shot at the information.

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/priority.go]: On closer inspection, this will also fix a bug with kubectl trying to submit a version of a kind the server doesn't have to a version the server does have

for example:
```
$ more role.yaml 
apiVersion: rbac.authorization.k8s.io/v2
kind: ClusterRole
...

kubectl create -f role.yaml --v=6
...
POST https://localhost:6443/apis/rbac.authorization.k8s.io/v1
...
Error from server (BadRequest): error when replacing "role.yaml": ClusterRole in version "v2" cannot be handled as a ClusterRole: no kind "ClusterRole" is registered for version "rbac.authorization.k8s.io/v2"
```

the client should not have attempted to send the v2 object to the v1 endpoint... it should have failed when looking up the rest mapping

File: staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper.go]: for example, this would improve the way we present the error while preserving well-structured data in the error:

```go
// NoKindMatchError is returned if the RESTMapper can't find any match for a kind
type NoKindMatchError struct {
	// GroupKind is the API group and kind that was searched
	GroupKind schema.GroupKind
	// SearchedVersions is the optional list of versions the search was restricted to
	SearchedVersions []string
}

func (e *NoKindMatchError) Error() string {
	searchedVersions := sets.NewString()
	for _, v := range e.SearchedVersions {
		searchedVersions.Insert(schema.GroupVersion{Group: e.GroupKind.Group, Version: v}.String())
	}

	switch len(searchedVersions) {
	case 0:
		return fmt.Sprintf("no matches for kind %q in group %q", e.GroupKind.Kind, e.GroupKind.Group)
	case 1:
		return fmt.Sprintf("no matches for kind %q in version %q", e.GroupKind.Kind, searchedVersions.List()[0])
	default:
		return fmt.Sprintf("no matches for kind %q in versions %q", e.GroupKind.Kind, searchedVersions.List())
	}
}
```

File: staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper_test.go [hunk 6]
```diff
@@ -261,42 +261,88 @@ func TestMultiRESTMapperRESTMappings(t *testing.T) {
...
+			name:      "return both",
+			mapper:    MultiRESTMapper{fixedRESTMapper{mappings: []*RESTMapping{mapping1}}, fixedRESTMapper{mappings: []*RESTMapping{mapping2}}},
+			groupKind: schema.GroupKind{Kind: "Foo"},
+			versions:  []string{"v1beta"},
+			result:    []*RESTMapping{mapping1, mapping2},
 		},
 	}
 
 	for _, tc := range tcs {
-		actualResult, actualErr := tc.mapper.RESTMappings(tc.input)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper_test.go]: can just do `actualResult, actualErr := tc.mapper.RESTMappings(tc.input, tc.versions...)`

**Comment by liggitt** [comment]
need to run `hack/update-bazel.sh`, one nit on the unit test, LGTM otherwise

**Comment by liggitt** [comment]
Still needs hack/update-bazel.sh run and the updated BUILD file committed

**Comment by chentao1596** [comment]
It's a strange thing, i have committed the BUILD, but the error is still exist. In fact, it's passed on my local host, and not any file should be committed exist to be found.

**Comment by liggitt** [comment]
Ah, this is the error now 

I0105 01:11:11.440] staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper_test.go:335:69: cannot use tc.versions (type []string) as type string in argument to tc.mapper.RESTMappings

File: staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper_test.go [hunk 7]
```diff
@@ -261,42 +261,78 @@ func TestMultiRESTMapperRESTMappings(t *testing.T) {
...
+		actualResult, actualErr := tc.mapper.RESTMappings(tc.groupKind, tc.versions)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper_test.go]: Should be `versions...`

Comment (chentao1596) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/multirestmapper_test.go]: oh, %>_<%, i will try to do and test it again ,thank you, ^_^

**Comment by chentao1596** [comment]
@liggitt Thank you for the patient inspection and guidance, test is passed! ^_^
